examples/sponsored-tx-enabled-auction/src/lib.rs

```diff
@@ -0,0 +1,433 @@
+    host.state_mut().counter = counter + 1;
+
+    // Insert the item into the state.
+    host.state_mut().items.insert(counter, ItemState {
+        auction_state:  AuctionState::NotSoldYet,
+        highest_bidder: None,
+        name:           item.name,
+        end:            item.end,
+        start:          item.start,
+        highest_bid:    item.minimum_bid,
+        creator:        sender_address,
+        token_id:       item.token_id,
+    });
+
+    Ok(())
+}
+
+/// The `bid` entry point in this contract is not meant to be invoked directly
+/// but rather through the `onCIS2Receive` hook mechanism in the cis2 token
+/// contract. Any account can bid for an item. The `bid` entry point can be
+/// invoked via a sponsored transaction mechanism (`permit` entry point) in case
+/// it is implemented in the cis2 token contract. The bidding flow starts with
+/// an account invoking either the `transfer` or the `permit` entry point in the
+/// cis2 token contract and including the `item_index` in the `additionalData`
+/// section of the input parameter. The `transfer` or the `permit` entry point
+/// will send some token amounts to this contract from the bidder. If the token
+/// amount exceeds the current highest bid, the bid is accepted and the previous
+/// highest bidder is refunded its token investment.
+#[receive(
+    contract = "sponsored_tx_enabled_auction",
+    name = "bid",
+    mutable,
+    parameter = "OnReceivingCis2Params<ContractTokenId, ContractTokenAmount>",
+    error = "Error"
+)]
+fn auction_bid(ctx: &ReceiveContext, host: &mut Host<State>) -> ReceiveResult<()> {
+    // Getting input parameters.
+    let params: OnReceivingCis2Params<ContractTokenId, ContractTokenAmount> =
+        ctx.parameter_cursor().get()?;
+
+    // Ensure the sender is the cis2 token contract.
+    if let Address::Contract(contract) = ctx.sender() {
+        ensure_eq!(contract, host.state().cis2_contract, Error::NotTokenContract.into());
+    } else {
+        bail!(Error::NotTokenContract.into())
+    };
+
+    // Get the item_index from the additionalData.
+    let additional_data_index: AdditionalDataIndex = from_bytes(params.data.as_ref())?;
+
+    let mut item = host
+        .state_mut()
+        .items
+        .entry(additional_data_index.item_index)
+        .occupied_or(Error::NoItem)?;
+
+    ensure_eq!(item.token_id, params.token_id, Error::WrongTokenID.into());
+
+    // Ensure that only accounts can bid for an item.
+    let bidder_address = match params.from {
+        Address::Contract(_) => bail!(Error::OnlyAccount.into()),
+        Address::Account(account_address) => account_address,
+    };
+
+    // Ensure the auction has not been finalized yet.
+    ensure_eq!(item.auction_state, AuctionState::NotSoldYet, Error::AuctionAlreadyFinalized.into());
+
+    let slot_time = ctx.metadata().slot_time();
+    // Ensure the auction has not ended yet.
+    ensure!(slot_time <= item.end, Error::BidTooLate.into());
```

Comment:
Consider moving these cheap checks up before the more expensive ones (such as parsing the parameter and looking up the item).

Reply:
I need to look up the item for this check as well, since I need to know `item.end` value.